### PR TITLE
Fix dropping namespace issue

### PR DESCRIPTION
**Pull request by chtyim**
- We should disable the stream table in Hive when dropping a stream
- Also fixed a annoying exception when StreamClient is used to fetch events.

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/FileStreamAdmin.java [hunk 1]
```diff
@@ -111,6 +111,7 @@ public void dropAllInNamespace(Id.Namespace namespace) throws Exception {
 
     for (final Location streamLocation : locations) {
       doTruncate(streamLocation);
+      alterExploreStream(StreamUtils.getStreamIdFromLocation(streamLocation), false);
```
Comment (jwang47) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/FileStreamAdmin.java]: Should this also be added to drop()? and then in here just call drop() instead of truncate and alterExploreStream?

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/FileStreamAdmin.java]: hum.. I thought it was done in drop...

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/FileStreamAdmin.java]: but apparently it is not

**Comment by bdmogal** [comment]
LGTM

**Comment by jwang47** [comment]
Can you add the following two test cases?

1. Test create/drop/create stream
  * After create stream: explore for stream should exist
2. Test create stream, delete ns, create ns, create stream
  * After first and list create stream, explore for stream should exist
  * After delete ns, explore for stream should not exist

I don't think (2) will pass right now..

File: cdap-data-fabric-tests/src/test/java/co/cask/cdap/data/stream/StreamFileJanitorTestBase.java [hunk 2]
```diff
@@ -126,15 +126,39 @@ public void testCleanupTTL() throws Exception {
...
+      }
+    } finally {
+      writer.close();
+    }
+
+    // Delete the stream
+    streamAdmin.drop(streamId);
+
+    // Run janitor. Should be running fine without exception.
+    janitor.cleanAll();
```
Comment (jwang47) [inline cdap-data-fabric-tests/src/test/java/co/cask/cdap/data/stream/StreamFileJanitorTestBase.java]: Check that stream no longer exists?

File: cdap-data-fabric-tests/src/test/java/co/cask/cdap/data2/transaction/stream/StreamAdminTest.java [hunk 3]
```diff
@@ -89,7 +91,7 @@ public void testDropAllInNamespace() throws Exception {
 
     // All of the streams within the default namespace should have no data in them
```
Comment (jwang47) [inline cdap-data-fabric-tests/src/test/java/co/cask/cdap/data2/transaction/stream/StreamAdminTest.java]: Update comment

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/FileStreamAdmin.java [hunk 4]
```diff
@@ -204,14 +204,26 @@ public void upgrade() throws Exception {
...
-    return loadConfig(streamLocation);
+    Location configLocation = getConfigLocation(streamId);
+    Preconditions.checkArgument(configLocation.exists(), "Stream '%s' does not exist.", streamId);
+
+    StreamConfig config = GSON.fromJson(
+      CharStreams.toString(CharStreams.newReaderSupplier(Locations.newInputSupplier(configLocation), Charsets.UTF_8)),
+      StreamConfig.class);
+
+    int threshold = config.getNotificationThresholdMB();
+    if (threshold <= 0) {
```
Comment (jwang47) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/FileStreamAdmin.java]: So the minimum notification threshold is 0? Is this a constant for this?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/FileStreamAdmin.java [hunk 5]
```diff
@@ -298,10 +304,9 @@ public StreamConfig call() throws Exception {
   }
 
   private void assertNamespaceHomeExists(Id.Namespace namespaceId) throws IOException {
-    Location namespaceHomeLocation = Locations.getParent(getStreamsHomeLocation(namespaceId));
-    Preconditions.checkArgument(namespaceHomeLocation.exists(),
-                                String.format("Home directory %s for namespace %s not found",
-                                              namespaceHomeLocation.toURI().getPath(), namespaceId));
+    Location namespaceHomeLocation = Locations.getParent(getNamespaceLocation(namespaceId));
+    Preconditions.checkArgument(namespaceHomeLocation == null || namespaceHomeLocation.exists(),
```
Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/FileStreamAdmin.java]: If namespaceHomeLocation == null, the exception won't be thrown.

File: cdap-data-fabric-tests/src/test/java/co/cask/cdap/data/stream/StreamFileJanitorTestBase.java (hunk 2 above)
Comment (chtyim) [inline cdap-data-fabric-tests/src/test/java/co/cask/cdap/data/stream/StreamFileJanitorTestBase.java]: There are other tests doing that already. In this Janitor test, it's mainly to verify the Janitor can handle the case correctly.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/FileStreamAdmin.java (hunk 4 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/FileStreamAdmin.java]: It's existing code. It is the default value of long field when deserializing using Gson and yet the field is missing.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/FileStreamAdmin.java (hunk 5 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/FileStreamAdmin.java]: good catch. I was thinking about it as a `if`. Precondition should flip the condition.

**Comment by chtyim** [comment]
Build passed: http://builds.cask.co/browse/CDAP-RBT193-10

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinatorClient.java [hunk 6]
```diff
@@ -81,6 +82,11 @@
    */
   protected abstract void streamCreated(Id.Stream streamId);
 
+  /**
+   * GEts invoked when a stream is deleted.
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinatorClient.java]: typo

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamFileJanitor.java [hunk 7]
```diff
@@ -115,4 +120,12 @@ void clean(StreamConfig config, long currentTime) throws IOException {
   private boolean isPartitionDirector(Location location) throws IOException {
     return (location.isDirectory() && location.getName().indexOf('.') > 0);
   }
+
+  private boolean isStreamExists(Id.Stream streamId) throws IOException {
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamFileJanitor.java]: Only for converting exceptions to IOException?

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamFileJanitor.java]: Yes, to simplify the try catch logic in the `clean` method

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/FileStreamAdmin.java [hunk 8]
```diff
@@ -326,42 +331,61 @@ private void createStreamFeeds(StreamConfig config) {
...
+   */
+  private Location getStreamLocation(Id.Stream streamId) throws IOException {
+    return getNamespaceLocation(streamId.getNamespace()).append(streamId.getName());
   }
 
-  private Location getStreamsHomeLocation(Id.Namespace namespace) throws IOException {
+  /**
+   * Returns the location for the given namespace that contains all streams belong to that namespace.
+   */
+  private Location getNamespaceLocation(Id.Namespace namespace) throws IOException {
```
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/FileStreamAdmin.java]: This might give a wrong idea, no? Its actually returning the streams directory under the namespace, and not the namespace location?

Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/FileStreamAdmin.java]: Maybe ``getStreamsLocationInNamespace``?

**Comment by jwang47** [comment]
LGTM

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/FileStreamAdmin.java (hunk 8 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/FileStreamAdmin.java]: It's a private method and it's inside the StreamAdmin. We don't need to name everything in a private method name as the class already gives the context that everything inside this class should be just operating on stream. Also, the javadoc says what it returns.

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/FileStreamAdmin.java]: I rename it to `getStreamBaseLocation`.

Comment (anwar6953) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/stream/FileStreamAdmin.java]: the method that takes in a streamId and returns: /streams/fooStream dir is also named similarly.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamFileJanitor.java (hunk 7 above)
Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamFileJanitor.java]: I see. Do we care about other exceptions that ``StreamAdmin.exists`` may throw in other parts of the code? If not maybe we can have this cleanup in ``StreamAdmin.exists`` itself? May be in a different PR though.

**Comment by bdmogal** [comment]
@chtyim only minor comments from me. LGTM.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamFileJanitor.java (hunk 7 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamFileJanitor.java]: `cleanAll` itself throws `IOException`. The caller needs to handle that. In production code, the caller is actually a executor thread that schedule to do cleanup and it just log the exception and schedule it again (hence if it's interim failure, it'll get retried when the janitor runs again in next scheduled time).

Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamFileJanitor.java]: Right. 

I was only wondering why ``StreamAdmin.exists`` should throw anything other than an ``IOException``. I just checked that ``HBaseQueueAdmin.exists`` throws a ``DatasetManagementException``, so that's probably why.

Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamFileJanitor.java]: It's a legacy issue. The `StreamAdmin` used to extend from `DatasetAdmin`, which throws `Exception` in all methods. When we detach `StreamAdmin` from `DatasetAdmin`, those method signature get retained.

Comment (bdmogal) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamFileJanitor.java]: I see. ok.
